browser/ui/webui/settings/brave_import_data_handler.cc

```diff
@@ -5,9 +5,168 @@
+#include "base/files/file_util.h"
+#include "base/mac/foundation_util.h"
+#include "base/task/thread_pool.h"
+#include "brave/browser/ui/webui/settings/brave_full_disk_access_confirm_dialog_delegate.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/browser/ui/tab_modal_confirm_dialog.h"
+#include "content/public/browser/browser_thread.h"
+#endif  // BUILDFLAG(IS_MAC)
+
+namespace {
+#if BUILDFLAG(IS_MAC)
+bool HasProperDiskAccessPermission(uint16_t imported_items) {
+  DCHECK(imported_items);
+
+  const base::FilePath& library_dir = base::mac::GetUserLibraryPath();
+  const base::FilePath safari_dir = library_dir.Append("Safari");
+
+  if (imported_items & importer::FAVORITES) {
+    const base::FilePath bookmarks_path = safari_dir.Append("Bookmarks.plist");
+    if (!PathIsWritable(bookmarks_path)) {
+      LOG(ERROR) << __func__ << " " << bookmarks_path << " is not accessible."
+                 << " Please check full disk access permission.";
+      return false;
+    }
+  }
+
+  if (imported_items & importer::HISTORY) {
+    const base::FilePath history_path = safari_dir.Append("History.plist");
+    if (!PathIsWritable(history_path)) {
+      LOG(ERROR) << __func__ << " " << history_path << " is not accessible."
+                 << " Please check full disk access permission.";
+      return false;
+    }
+  }
+
+  return true;
+}
+#endif  // BUILDFLAG(IS_MAC)
+}  // namespace
+
 namespace settings {
 
 BraveImportDataHandler::BraveImportDataHandler() = default;
 BraveImportDataHandler::~BraveImportDataHandler() = default;
 
+void BraveImportDataHandler::StartImport(
+    const importer::SourceProfile& source_profile,
+    uint16_t imported_items) {
+#if BUILDFLAG(IS_MAC)
+  CheckDiskAccess(source_profile, imported_items);
+#else
+  StartImportImpl(source_profile, imported_items);
+#endif
+}
+
+void BraveImportDataHandler::StartImportImpl(
+    const importer::SourceProfile& source_profile,
+    uint16_t imported_items) {
+  // Temporary flag to keep old way until
+  // https://github.com/brave/brave-core/pull/15637 landed.
+  // Should be removed in that PR.
+  if (!IsParallelImportEnabled()) {
+    ImportDataHandler::StartImport(source_profile, imported_items);
+    return;
+  }
+  if (!imported_items)
+    return;
+  // If another import is already ongoing, let it finish silently.
+  if (import_observers_.count(source_profile.source_path))
+    import_observers_.erase(source_profile.source_path);
+
+  // Using weak pointers because it destoys itself when finshed.
+  auto* importer_host = new ExternalProcessImporterHost();
+  import_observers_[source_profile.source_path] =
+      std::make_unique<settings::BraveImporterObserver>(
+          importer_host, source_profile, imported_items,
+          base::BindRepeating(&BraveImportDataHandler::NotifyImportProgress,
+                              weak_factory_.GetWeakPtr()));
+  importer_host->set_observer(
+      import_observers_[source_profile.source_path].get());
```

Comment:
How about moving this to `BraveImporterObserver`'s ctor?
I think it would be betther set/reset observer by same instance.

Reply:
done